File: backend/src/core/ephemeris/interface.ts
import type {
  DateTime,
  CelestialBody,
  LunarDay,
  VoidOfCourseMoon,
  Aspect,
  House,
  PlanetaryHour,
  PlanetsApiResponse,
  AspectsApiResponse,
  AspectPerfectionsApiResponse,
  HousesApiResponse,
  VoidMoonApiResponse,
  PlanetaryHoursApiResponse,
  RetrogradesApiResponse,
  DispositorChainsResponse,
} from '@adaptive-astro/shared/types/astrology';

/**
 * Ephemeris Calculator Interface
 *
 * This interface defines the contract for astronomical calculations.
 * Implementations must provide precise planetary positions, lunar data,
 * and astrological calculations.
 */
export interface IEphemerisCalculator {
  /**
   * Get positions of all major planets for a given date/time
   * @param dateTime - Date, time, and location
   * @returns Planet positions from API
   */
  getPlanetsPositions(dateTime: DateTime): Promise<PlanetsApiResponse>;

  /**
   * Get aspects between planets for a given date/time
   * @param dateTime - Date and time
   * @param orb - Orb tolerance (default: 8 degrees)
   * @returns Planetary aspects from API
   */
  getAspects(dateTime: DateTime, orb?: number): Promise<AspectsApiResponse>;

  /**
   * Find every exact aspect perfection between the given planet pairs and
   * aspect types within [start, end). Each entry's `exact_at` is the UTC
   * moment the aspect becomes exact, accurate to ~1 minute.
   *
   * Used by v2 optimal-timing to answer "did aspect X perfect on day D?"
   * — a stronger signal than "is X applying at noon?".
   *
   * @param start ISO 8601 timestamp (UTC) — window start (inclusive)
   * @param end   ISO 8601 timestamp (UTC) — window end (exclusive)
   * @param pairs Planet pairs, e.g. [['Moon','Jupiter']]
   * @param aspects Aspect names, e.g. ['trine','sextile','conjunction']
   */
  getAspectPerfections(
    start: string,
    end: string,
    pairs: ReadonlyArray<readonly [string, string]>,
    aspects: ReadonlyArray<string>,
  ): Promise<AspectPerfectionsApiResponse>;

  /**
   * Get astrological houses for a given date/time/location
   * @param dateTime - Date, time, and location
   * @param system - House system ('placidus', 'whole-sign', 'equal')
   * @returns Houses from API
   */
  getHouses(dateTime: DateTime, system?: string): Promise<HousesApiResponse>;

  /**
   * Get planetary hours for a given date and location
   * @param dateTime - Date and location
   * @returns Planetary hours from API
   */
  getPlanetaryHours(dateTime: DateTime): Promise<PlanetaryHoursApiResponse>;

  /**
   * Get retrograde planets for a given date
   * @param dateTime - Date to check
   * @returns Retrograde planets from API
   */
  getRetrogradePlanets(dateTime: DateTime): Promise<RetrogradesApiResponse>;

  /**
   * Get current moon phase (illumination percentage)
   * @param dateTime - Date and time
   * @returns Moon illumination (0.0 = new moon, 1.0 = full moon)
   */
  getMoonPhase(dateTime: DateTime): Promise<number>;

  /**
   * Calculate lunar day (1-30) for given date/time
   * @param dateTime - Date and time
   * @returns Lunar day number and metadata
   */
  getLunarDay(dateTime: DateTime): Promise<LunarDay>;

  /**
   * Detect Void of Course Moon periods
   * @param dateTime - Date and time to check
   * @returns VoidOfCourseMoon data from API
   */
  getVoidOfCourseMoon(dateTime: DateTime): Promise<VoidMoonApiResponse>;

  /**
   * Get dispositor chains — who rules the ruler?
   * @param dateTime - Date and time
   * @param system - 'traditional' | 'modern' (default: 'traditional')
   */
  getDispositorChains(dateTime: DateTime, system?: string): Promise<DispositorChainsResponse>;
}

/**
 * Planet Positions Collection
 */
export interface PlanetPositions {
  sun: CelestialBody;
  moon: CelestialBody;
  mercury: CelestialBody;
  venus: CelestialBody;
  mars: CelestialBody;
  jupiter: CelestialBody;
  saturn: CelestialBody;
  uranus: CelestialBody;
  neptune: CelestialBody;
  pluto: CelestialBody;
}

File: lunar-calendar-api/app/core/ephemeris/interface.py
"""
Ephemeris Calculator Interface

Defines the contract that all ephemeris calculators must implement.
This allows for different implementations (Skyfield, Swiss Ephemeris, API-based)
while maintaining a consistent interface.
"""

from abc import ABC, abstractmethod
from typing import List, Optional, Dict
from datetime import datetime

from .types import (
    DateTime,
    PlanetPositions,
    MoonPhase,
    LunarDay,
    VoidOfCourseMoon,
    CelestialBody,
    Aspect,
    House,
    HouseSystem,
    PlanetaryHour,
    PlanetName,
)


class IEphemerisCalculator(ABC):
    """
    Interface for ephemeris calculations.

    All implementations must provide these methods with the specified signatures.
    """

    @abstractmethod
    async def get_planets_positions(self, date_time: DateTime) -> PlanetPositions:
        """
        Get positions of all major planets for a given date/time.

        Args:
            date_time: Date, time, and location

        Returns:
            Object with all planet positions

        Raises:
            EphemerisError: If calculation fails
        """
        pass

    @abstractmethod
    async def get_moon_phase(self, date_time: DateTime) -> MoonPhase:
        """
        Get current moon phase information.

        Args:
            date_time: Date and time

        Returns:
            Moon phase information including illumination percentage

        Raises:
            EphemerisError: If calculation fails
        """
        pass

    @abstractmethod
    async def get_lunar_day(self, date_time: DateTime) -> LunarDay:
        """
        Calculate lunar day (1-30) for given date/time.

        Args:
            date_time: Date and time

        Returns:
            Lunar day number and metadata

        Raises:
            EphemerisError: If calculation fails
        """
        pass

    @abstractmethod
    async def get_void_of_course_moon(
        self,
        date_time: DateTime
    ) -> Optional[VoidOfCourseMoon]:
        """
        Detect Void of Course Moon periods.

        The Moon is Void of Course between its last major aspect in a sign
        and the time it enters the next sign.

        Args:
            date_time: Date and time to check

        Returns:
            VoidOfCourseMoon object if active, None otherwise

        Raises:
            EphemerisError: If calculation fails
        """
        pass

    @abstractmethod
    async def get_retrograde_planets(
        self,
        date_time: DateTime
    ) -> List[CelestialBody]:
        """
        Get list of retrograde planets.

        Args:
            date_time: Date and time

        Returns:
            Array of planets currently in retrograde

        Raises:
            EphemerisError: If calculation fails
        """
        pass

    @abstractmethod
    async def calculate_aspects(
        self,
        bodies: List[CelestialBody],
        orb: Optional[float] = None
    ) -> List[Aspect]:
        """
        Calculate aspects between celestial bodies.

        Args:
            bodies: Array of celestial bodies to check
            orb: Orb tolerance in degrees (uses defaults if not specified)

        Returns:
            Array of aspects found

        Raises:
            EphemerisError: If calculation fails
        """
        pass

    @abstractmethod
    async def calculate_houses(
        self,
        date_time: DateTime,
        system: HouseSystem = HouseSystem.PLACIDUS
    ) -> Dict[int, House]:
        """
        Calculate house cusps for given birth data.

        Args:
            date_time: Birth date, time, and location
            system: House system to use (default: Placidus)

        Returns:
            Dictionary mapping house numbers (1-12) to House objects

        Raises:
            EphemerisError: If calculation fails
        """
        pass

    @abstractmethod
    async def get_planetary_hours(
        self,
        date_time: DateTime
    ) -> List[PlanetaryHour]:
        """
        Get planetary hours for a given date and location.

        Planetary hours divide the day and night into 12 hours each,
        with each hour ruled by a planet.

        Args:
            date_time: Date and location

        Returns:
            Array of planetary hours for the day

        Raises:
            EphemerisError: If calculation fails
        """
        pass

    # Optional helper methods

    async def get_planet_position(
        self,
        planet: PlanetName,
        date_time: DateTime
    ) -> CelestialBody:
        """
        Get position of a single planet.

        Default implementation calls get_planets_positions and extracts one planet.
        Can be overridden for optimization.

        Args:
            planet: Planet to get position for
            date_time: Date and time

        Returns:
            Position of the specified planet

        Raises:
            EphemerisError: If calculation fails
        """
        positions = await self.get_planets_positions(date_time)
        return positions.get_planet(planet)

    async def find_aspects_for_planet(
        self,
        planet: CelestialBody,
        other_planets: List[CelestialBody],
        orb: Optional[float] = None
    ) -> List[Aspect]:
        """
        Find all aspects for a specific planet.

        Args:
            planet: The planet to find aspects for
            other_planets: Other planets to check aspects with
            orb: Orb tolerance in degrees

        Returns:
            List of aspects involving the specified planet
        """
        all_bodies = [planet] + other_planets
        all_aspects = await self.calculate_aspects(all_bodies, orb)
        return [
            aspect for aspect in all_aspects
            if aspect.body1.name == planet.name or aspect.body2.name == planet.name
        ]
